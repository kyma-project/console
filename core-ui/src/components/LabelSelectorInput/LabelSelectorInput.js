import React from 'react';
import './LabelSelectorInput.scss';
import { Token } from 'fundamental-react/Token';

//TODO: move this component to a shared "place"

const domainSegmentRegexp = '([a-z0-9]([a-z0-9-_]{0,61}[a-z0-9])?)';

// Dot needs to be escaped for regexp
const domainRegexp = `(${domainSegmentRegexp}\\.)*${domainSegmentRegexp}`;
const nameAndValueRegexp = '[a-z0-9A-Z]([a-z0-9A-Z-_\\.]{0,61}[a-z0-9A-Z])?';
export const labelRegexp = new RegExp(
  `^((${domainRegexp})/)?${nameAndValueRegexp}=(${nameAndValueRegexp})?$`,
);

export const Label = ({ text, onClick }) => (
  <Token
    title="Click to remove"
    className="label-selector__label"
    onClick={onClick}
  >
    {text}
  </Token>
);

export const NonRemovableLabel = ({ text }) => (
  <Token className="label-selector__label--non-removable">{text}</Token>
);

const LabelSelectorInput = ({ labels = [], readonlyLabels = [], onChange }) => {
  function handleLabelEntered(e) {
    const value = e.target.value;
    if (e.keyCode !== 13 || !labelRegexp.test(value)) return;
    e.target.value = '';
    onChange([...labels, value]);
  }
  function handleLabelRemoved(label) {
    onChange(labels.filter(l => l !== label));
  }

  return (
    <div className="fd-form__set">
      <div className="label-selector">
        {readonlyLabels.map(l => (
          <NonRemovableLabel key={l} text={l} />
        ))}

        {labels.map(l => (
          <Label key={l} text={l} onClick={() => handleLabelRemoved(l)} />
        ))}
        <input
          className="fd-form__control label-selector__input"
          type="text"
          placeholder="Enter new label"
          onKeyDown={handleLabelEntered}
        />
      </div>
    </div>
  );
};

export default LabelSelectorInput;
