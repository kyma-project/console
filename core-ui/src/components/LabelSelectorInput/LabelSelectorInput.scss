.label-selector {
  color: var(--fd-forms-color);
  outline: 0;

  border-style: solid;
  border-width: 1px;
  border-color: var(--fd-forms-border-color);
  background-color: var(--fd-forms-background-color);
  border-radius: 4px;
  transition: border-color 125ms;

  box-sizing: border-box;
  min-height: var(--fd-forms-height);
  padding-left: 8px;

  display: flex;
  align-items: center;
  flex-wrap: wrap;

  &:active,
  &:focus {
    box-shadow: 0 0 0 1px var(--fd-color-action-focus);
  }

  input {
    height: calc(var(--fd-forms-height) - 2px);
    width: auto;
    padding-left: 8px;

    border: none;
    flex-grow: 1;

    &:active,
    &:focus {
      box-shadow: none !important;
    }
  }

  &__label {
    height: calc(var(--fd-forms-height) - 2 * 6px);
    margin: 4px 0;
    flex-shrink: 0;
    border: 1px solid var(--fd-color-neutral-3);

    &:not(:last-of-type) {
      margin-right: 4px;
    }

    &--non-removable {
      @extend .label-selector__label;
      cursor: default;

      &:after {
        content: unset;
      }
    }
  }
}
