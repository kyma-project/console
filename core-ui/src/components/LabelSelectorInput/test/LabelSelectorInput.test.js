import React from 'react';
import LabelSelectorInput, { labelRegexp } from '../LabelSelectorInput';
import { Label, NonRemovableLabel } from '../LabelSelectorInput';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import { descriptions } from 'jest-config';

describe('LabelSelectorInput', () => {
  const mockChange = jest.fn();
  afterEach(() => {
    mockChange.mockReset();
  });
  it('Renders with minimal props', () => {
    const component = renderer.create(<LabelSelectorInput />);
    expect(component).toBeTruthy();
  });

  it('Renders readonly labels', () => {
    const component = shallow(
      <LabelSelectorInput readonlyLabels={['a', 'b']} />,
    );
    expect(component.find(NonRemovableLabel)).toMatchSnapshot();
  });

  it('Renders labels', () => {
    const component = shallow(<LabelSelectorInput labels={['a', 'b']} />);
    expect(component.find(Label)).toMatchSnapshot();
  });

  it(`Doesn't fire onChange with invalid label`, () => {
    const component = mount(<LabelSelectorInput onChange={mockChange} />);
    const input = component.find('input');

    input.instance().value = 'abc';
    input.simulate('keydown', { keyCode: 13 });

    expect(mockChange.mock.calls.length).toBe(0);
  });

  it(`Fires onChange when valid label is entered`, () => {
    const component = mount(<LabelSelectorInput onChange={mockChange} />);
    const input = component.find('input');

    input.instance().value = 'abc=def';
    input.simulate('keydown', { keyCode: 13 });

    expect(mockChange.mock.calls).toMatchSnapshot();
  });

  it(`Allows to remove labels`, () => {
    const component = mount(
      <LabelSelectorInput labels={['a=b', 'c=d']} onChange={mockChange} />,
    );

    component
      .find(Label)
      .first()
      .simulate('click');

    expect(mockChange.mock.calls).toMatchSnapshot();
  });

  it(`Doesn' allow to remove readonly labels`, () => {
    const component = mount(
      <LabelSelectorInput
        readonlyLabels={['a=b', 'c=d']}
        onChange={mockChange}
      />,
    );

    component
      .find(NonRemovableLabel)
      .first()
      .simulate('click');

    expect(mockChange.mock.calls.length).toBe(0);
  });
});
