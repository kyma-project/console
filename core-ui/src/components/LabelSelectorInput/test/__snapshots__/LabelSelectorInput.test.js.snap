// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabelSelectorInput Allows to remove labels 1`] = `
Array [
  Array [
    Array [
      "c=d",
    ],
  ],
]
`;

exports[`LabelSelectorInput Fires onChange when valid label is entered 1`] = `
Array [
  Array [
    Array [
      "abc=def",
    ],
  ],
]
`;

exports[`LabelSelectorInput Renders labels 1`] = `
Array [
  <Label
    key="a"
    onClick={[Function]}
    text="a"
  />,
  <Label
    key="b"
    onClick={[Function]}
    text="b"
  />,
]
`;

exports[`LabelSelectorInput Renders readonly labels 1`] = `
Array [
  <NonRemovableLabel
    key="a"
    text="a"
  />,
  <NonRemovableLabel
    key="b"
    text="b"
  />,
]
`;
